querier.go

```diff
@@ -266,6 +267,62 @@ func (q *blockQuerier) Close() error {
 	return merr.Err()
 }
 
+// Bitmap used by func special to check whether a character needs to be escaped.
+var specialBytes [16]byte
+
+// special reports whether byte b needs to be escaped.
+func special(b byte) bool {
+	return b < utf8.RuneSelf && specialBytes[b%16]&(1<<(b/16)) != 0
+}
+
+func init() {
+	for _, b := range []byte(`.+*?()|[]{}^$`) {
+		specialBytes[b%16] |= 1 << (b / 16)
+	}
+}
+
+func findSetMatches(pattern string) []string {
+	// Return empty matches if the wrapper from Prometheus is missing.
+	if len(pattern) < 6 || pattern[:4] != "^(?:" || pattern[len(pattern)-2:] != ")$" {
+		return []string{}
+	}
+	escaped := false
+	sets := []*strings.Builder{&strings.Builder{}}
+	for i := 4; i < len(pattern)-2; i++ {
+		if escaped {
+			// Add the escaped special character to the sets.
+			if special(pattern[i]) {
+				sets[len(sets)-1].WriteByte(pattern[i])
+			} else if pattern[i] == '\\' {
+				sets[len(sets)-1].WriteByte('\\')
+			} else {
+				return []string{}
+			}
+			escaped = false
+		} else {
+			// Return empty sets when there are special characters excluding '|'.
+			if special(pattern[i]) {
+				if pattern[i] == '|' {
+					sets = append(sets, &strings.Builder{})
+				} else {
+					return []string{}
+				}
+			} else if pattern[i] == '\\' {
+				escaped = true
+			} else {
+				sets[len(sets)-1].WriteByte(pattern[i])
+			}
```

Comment:
(minor personal opinion) IMO case differentiations like this are visually easier to follow using `switch`:

```go
switch {
case special(pattern[i]):
  // ...
case pattern[i] == '\\':
  // ...
default:
  // ...
}
```